### ISSUE TITLE
DirRead autocmd

### ISSUE THREAD

**Issue by Shougo** [vim/vim#16259]
`BufRead` autocmd does not be fired when directory is opened.
It is hard to implement file explore plugins to detect directory is opened.
So I think new autocmd `DirRead` is needed.

Test commands

```vim
autocmd BufRead * echomsg expand('<afile>')
edit .
```

Note: "`.` is directory" message is only printed.

**Comment by Shougo** [comment]
For example, netrw uses other autocmds.

```vim
augroup FileExplorer
 au!
 au BufLeave *  if &ft != "netrw"|let w:netrw_prvfile= expand("%:p")|endif
 au BufEnter *	sil call s:LocalBrowse(expand("<amatch>"))
 au VimEnter *	sil call s:VimEnter(expand("<amatch>"))
 if has("win32")
  au BufEnter .* sil call s:LocalBrowse(expand("<amatch>"))
 endif
augroup END
```

I think it is hack.

**Comment by justinmk** [comment]
Only a `BufEnter` handler is needed, 

    autocmd BufEnter * if <SID>isdir(expand('%:p')) ...

but for reference, the Dirvish `<SID>isdir()` function to decide whether a file is representing a "directory" is definitely hacky: https://github.com/justinmk/vim-dirvish/blob/c87a8e103f2e8b44b411a38dda9bbe1229547952/plugin/dirvish.vim#L9-L15

delegating that decision to the editor would be nice. That could be achieved by treating `isdirectory(f:number)` differently than `isdirectory(f:string)`, or add an optional parameter like `isdirectory(f:any, isbuf?:boolean)`.

Note also that directory browser plugins historically must clear netrw's handler: https://github.com/justinmk/vim-dirvish/blob/c87a8e103f2e8b44b411a38dda9bbe1229547952/plugin/dirvish.vim#L20-L21

    autocmd VimEnter * if exists('#FileExplorer') | exe 'au! FileExplorer *' | endif

So a more formal way of declaring which plugin "owns" directory browsing could be useful.

**Comment by Shougo** [comment]
> Only a BufEnter handler is needed,

It is possible.  But `BufEnter` is more general autocmd.

> Note also that directory browser plugins historically must clear netrw's handler: https://github.com/justinmk/vim-dirvish/blob/c87a8e103f2e8b44b411a38dda9bbe1229547952/plugin/dirvish.vim#L20-L21

I know...  I think netrw uses `DirRead` autocmd for it, other file explorer can remove the event more easily.

**Comment by Shougo** [comment]
> As suggested in https://github.com/neovim/neovim/issues/31621 , it does seem attractive for vim to define a directory 'filetype'. That would allow all the usual mechanisms to be used, such as ftplugins.

If Vim/Neovim implements that, I think they should add a new buffer type(`buftype`) for directories instead. `filetype` is already used for various features, so making it represent directories would break the compatibility.

**Comment by justinmk** [comment]
> `filetype` is already used for various features, so making it represent directories would break the compatibility.

? Dirvish defines a `dirvish` filetype for its directory buffers. Why would defining a `directory` filetype break compatibility?

**Comment by Shougo** [comment]
> ? Dirvish defines a dirvish filetype for its directory buffers. Why would defining a directory filetype break compatibility?

Each file explorer plugin sets its own filetype, so it cannot be unified. Even if unified, there is a problem that users cannot distinguish between plugins.

**Comment by justinmk** [comment]
> Each file explorer plugin sets its own filetype, so it cannot be unified.

Only one plugin can control dir buffers. They don't work together.

> Even if unified, there is a problem that users cannot distinguish between plugins.

There can only be one dir browser plugin, they don't work together. So users can just check for `filetype=='directory'`, knowing that whichever dir browser plugin they installed is the one that will be active.

**Comment by Shougo** [comment]
> There can only be one dir browser plugin, they don't work together. So users can just check for filetype=='directory', knowing that whichever dir browser plugin they installed is the one that will be active.

For example, current file explorer plugin uses the same filetype for both real directory and buffers.
If netrw uses `directory` filetype for real directory, other netrw buffers use different filetype?
I think it will be confused.

**Comment by Shougo** [comment]
And users know what is the real plugin name.  But from other plugins?

Some plugins change their behavior depending on which plugin is active. It is difficult to distinguish them if they have the same filetype.

**Comment by justinmk** [comment]
> If netrw uses `directory` filetype for real directory, other netrw buffers use different filetype?
> I think it will be confused.

No, it's confusing that netrw already *doesn't* use a different filetype for non-directory buffers. That's just wrong.

> And users know what is the real plugin name. But from other plugins?

This is how all filetypes work.

**Comment by LunarLambda** [comment]
I do think it would help the plugin ecosystem significantly if we could have a single, agreed on hookpoint for installing a 'directory handler' plugin of sorts.

Right now, plugins like netrw or oil.nvim have to hook into BufAdd or BufReadPre, check if the buffer is a directory, then either hook further events such as BufReadCmd, or rename the buffer to some internal name (e.g., `oil:///path`) to then correctly initialize things. This means the plugin has to listen to far more events with a far broader filter than it actually cares about, and means every file browser implements its own checking logic, often in ways which are subtly different in behaviour.

Furthermore, plugins are required to check `VimEnter` as buffers opened from the command line are created before plugins are able to register autocommands for them.

This means that plugins are currently forced to handle 2-3 different ways of opening a directory, and many different edge cases.

A directory-related autocommand could theoretically impose the 'only one plugin gets to be the default file browser' restriction by only allowing a single handler to be registered, and requiring `:au!` to replace/delete it, and making the various `Buf`/`File` events *consistently* not trigger for directories. Then other plugins could still be used via custom commands, bindings, or "protocol style" autocommand handlers.

Ideally, we'd also figure out a way to avoid the 'need to handle VimEnter' issue. Maybe buffers opened from the command line/before plugins are loaded only fire a "DirectoryRead" event once a plugin has registered a handler for it?

It's maybe a lot of special behaviour to ascribe to one event, but I think it's already the case that different autocommands have at times wildly different restrictions or quirks (such as `<amatch>` vs `<afile>`, etc.)

**Comment by hoofcushion** [comment]
> This means that plugins are currently forced to handle 2-3 different ways of opening a directory, and many different edge cases

Not only directory, so does a nomal file or filetype, a unified way of waken plugin attach on different file/dir/filetype on startup/runtime
I currently using VimEnter and BufAdd and checking `vim.uv.fs_stat(ev.file)~=nil` (since) to deterime if buf is a file, after that a BufNewFile is needed for detecting file open by args. which is quite anoying at the first place because I had tried many combination of autocmd and test it if's working on these edge cases.
And so does filetypes, after creating attach autocmd, I have to manually check if currently opened buffer match the filetypes, else it will miss.
There is too may edge cases for defining buffer attach stratagy in vim.

**Comment by Shougo** [comment]
Yes.  I think file handler autocmd is needed.

**Comment by hoofcushion** [comment]
According to my summary, plugin base on file may need to set these hook to load properly:
- BufNewFile
- BufAdd
- VimEnter
- SessionLoadPost (ignored by many buffer related autocmd)
- Check directory buffer
- Check current buffer
And block these:
- any buffer that buftype is not ""
The problem is that these buffer-related checks are too granular, causing many edge cases, adding more specific autocmd brings even more complexity,

**Comment by matveyt** [comment]
That's overkill. All one has to do is to trap `BufNew` and if it's `isdirectory()` then set up `BufReadCmd` for this particular buffer. No need to add any extra autocommands.

If anyone is interested the code example is [here](https://github.com/matveyt/vim-drvo/blob/74ad4fe418c68a6a8e6876f73891c6fff831c4e0/plugin/drvo.vim#L15).

**Comment by Shougo** [comment]
> That's overkill. All one has to do is to trap BufNew and if it's isdirectory() then set up BufReadCmd for this particular buffer. No need to add any extra autocommands.

I have checked it.  `BufNew` seems works.  But `VimEnter` also needed.  Because `BufNew` is not fired when starting Vim.
But why `netrw` uses `BufEnter`?  It is wrong implementation?

**Comment by matveyt** [comment]
@Shougo No `VimEnter`. It is enough to iterate through all pre-loaded buffers once on plugin startup.
I've no idea why everyone likes to pollute `BufEnter`. In my implementation I go without it.

**Comment by Shougo** [comment]
> It is enough to iterate through all pre-loaded buffers once on plugin startup.

I see. That indeed seems sufficient.

The biggest problem with BufNew is that the current buffer(and also window?) might be different from the buffer that was created. However, this is an issue that can be prevented by implementing the plugin carefully.

**Comment by matveyt** [comment]
>The biggest problem with BufNew is that the current buffer(and also window?) might be different from the buffer that was created.

This is normal for some autocommands including `BufNew`. Yet we're going to postpone anything, except `isdirectory()` check, until `BufReadCmd` anyway. And `BufReadCmd` is actually triggered for active buffer.

>Were there any implementation issues that required using BufEnter?

I mentioned window-locals related issue above. Don't know about anything else.
